lib/Sema/CSFix.cpp

```diff
@@ -871,6 +871,47 @@ IgnoreContextualType *IgnoreContextualType::create(ConstraintSystem &cs,
       IgnoreContextualType(cs, resultTy, specifiedTy, locator);
 }
 
+bool RemoveUnnecessaryCoercion::diagnose(bool asNote) const {
+  auto &cs = getConstraintSystem();
+  UnnecessaryCoercionFailure failure(cs, getFromType(), getToType(),
+                                     getLocator());
+  return failure.diagnose(asNote);
+}
+
+bool RemoveUnnecessaryCoercion::attempt(ConstraintSystem &cs, Type fromType,
+                                        Type toType,
+                                        ConstraintLocatorBuilder locator) {
+  auto last = locator.last();
+  bool isExplicitCoercion =
+      last && last->is<LocatorPathElt::ExplicitTypeCoercion>();
+  if (!isExplicitCoercion)
+    return false;
+
+  auto expr = cast<CoerceExpr>(locator.getAnchor());
+  auto toTypeRepr = expr->getCastTypeLoc().getTypeRepr();
+
+  // Only diagnosing for coercion where the left-side is a DeclRefExpr
+  // or a explicit/implicit coercion e.g. Double(1) as Double
+  if (!isa<ImplicitlyUnwrappedOptionalTypeRepr>(toTypeRepr) &&
+      (isa<DeclRefExpr>(expr->getSubExpr()) ||
+       isa<CoerceExpr>(expr->getSubExpr()))) {
+    auto coercedType = cs.getType(expr->getSubExpr());
+    if (auto *typeVariable = coercedType->getAs<TypeVariableType>()) {
+      if (auto *binding = cs.getCurrentTypeVariableBinding(typeVariable)) {
+        // The type variable binding comes from the coercion, so we skip it.
+        if (binding->getBindingSource() >= ConstraintKind::Conversion)
```

Comment:
Check for conversion here is doesn't seem to be sufficient to determine whether binding comes from _this_ coercion, it only determines that binding comes from _some_ conversion constraint...